src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -303,111 +299,138 @@ void Transceiver::process_tasks_() {
                 packet::address_to_str(*task.address).c_str());
+
         return false;
     }
 
-    core::SharedPtr<UDPReceiver> rp = new (allocator_)
-        UDPReceiver(loop_, *task.writer, packet_pool_, buffer_pool_, allocator_);
+    return true;
+}
 
-    if (!rp) {
-        roc_log(LogError, "transceiver: can't add port %s: can't allocate receiver",
-                packet::address_to_str(*task.address).c_str());
+bool Transceiver::add_udp_sender_(Task& task) {
+    roc_panic_if(stopped_);
+
+    core::SharedPtr<UDPSender> sp = make_udp_sender_(task);
+    if (!sp) {
         return false;
     }
 
-    if (!rp->start(*task.address)) {
-        roc_log(LogError, "transceiver: can't add port %s: can't start receiver",
+    task.element = sp.get();
+    elements_.push_back(*sp);
+
+    if (!sp->start(*task.address)) {
+        roc_log(LogError, "transceiver: can't add port %s: can't start sender",
                 packet::address_to_str(*task.address).c_str());
+
         return false;
     }
 
-    receivers_.push_back(*rp);
-    num_ports_++;
+    task.writer = sp.get();
 
     return true;
 }
 
-bool Transceiver::add_udp_sender_(Task& task) {
-    if (stopped_) {
-        roc_log(LogError, "transceiver: can't add port %s: transceiver is stopped",
+core::SharedPtr<UDPReceiver> Transceiver::make_udp_receiver_(Task& task) {
+    if (has_port_(*task.address)) {
+        roc_log(LogError, "transceiver: can't add port %s: duplicate address",
                 packet::address_to_str(*task.address).c_str());
-        return false;
+
+        return NULL;
     }
 
+    core::SharedPtr<UDPReceiver> rp = new (allocator_)
+        UDPReceiver(*this, loop_, *task.writer, packet_pool_, buffer_pool_, allocator_);
+
+    if (!rp) {
+        roc_log(LogError, "transceiver: can't add port %s: can't allocate receiver",
+                packet::address_to_str(*task.address).c_str());
+
+        return NULL;
+    }
+
+    return rp;
+}
+
+core::SharedPtr<UDPSender> Transceiver::make_udp_sender_(Task& task) {
     if (has_port_(*task.address)) {
         roc_log(LogError, "transceiver: can't add port %s: duplicate address",
                 packet::address_to_str(*task.address).c_str());
-        return false;
+
+        return NULL;
     }
 
-    core::SharedPtr<UDPSender> sp = new (allocator_) UDPSender(loop_, allocator_);
+    core::SharedPtr<UDPSender> sp = new (allocator_) UDPSender(*this, loop_, allocator_);
 
     if (!sp) {
         roc_log(LogError, "transceiver: can't add port %s: can't allocate sender",
                 packet::address_to_str(*task.address).c_str());
-        return false;
+
+        return NULL;
     }
 
-    if (!sp->start(*task.address)) {
-        roc_log(LogError, "transceiver: can't add port %s: can't start sender",
-                packet::address_to_str(*task.address).c_str());
+    return sp;
+}
+
+bool Transceiver::has_port_(const packet::Address& address) const {
+    return get_element_(address) != NULL;
```

Comment:
I thinks we can remove this method (inline in into two invocation points).

Reply:
BTW, do we really need this check in make_udp_receiver_() and make_udp_sender_()?

Reply:
Reworked